### PR TITLE
the stack chapter

### LINKED ISSUE #16: Write a Chapter: "Framework & Library Choices"

### PR DESCRIPTION
**Pull request by Pomax**
fixes #16

### CODE REVIEW

File: app/_posts/2014-12-03-the-stack.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+## Service endpoints without client UI
+
+- [Node.js](http://nodejs.org)
+- [Hapi](http://hapijs.com)
+
+## User-facing apps
+
+Mobile first, including taking into account low-fi, no-fi, or dollars per megabytes, and localization.
+
+### Server side
```
Comment (Pomax) [inline app/_posts/2014-12-03-the-stack.md]: needs a section on data storage (nosql/sql) and associated migration libraries

**Comment by Pomax** [comment]
After 2 days of evaluating Hapi+Joi, I can only conclude that they're *seriously* horrible to learn (the official tutorials and docs are terrible) and I don't want to subject anyone else in our team to that when Express does what we need, so I'm rewriting the chapter to reflect that. Maybe once Hapi and Joi have better docs, intake, and APIs (the Joi API is truly dire) we can reassess them. For now, I'm calling veto on them.

File: app/_posts/2014-12-03-the-stack.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+- [Node.js](http://nodejs.org)
+- [Express](http://expressjs.com)
+- `npm` for dependencies
+
+Ideally, no server-side templating (ejs, nunjucks).
+
+- Data stores: SQL/NoSQL?
```
Comment (Pomax) [inline app/_posts/2014-12-03-the-stack.md]: a decision still needs to be made on which storage technology/technologies to go with.

**Comment by k88hudson** [comment]
I also feel like we should consider exploring Hapi – see

http://hueniverse.com/2012/12/20/hapi-a-prologue/

and 

http://matt-harrison.com/moving-from-express-to-hapi-js/

Where I think Hapi gets really interesting is testing. See https://medium.com/the-spumko-suite/testing-hapi-services-with-lab-96ac463c490a

**Comment by k88hudson** [comment]
I think testing is one of our weakest areas as a team – we need structure. If we *do* choose to stick with express, we need to adjust/define the way we build express modules/routes/etc. so they are testable since express is non-opinionated about it

**Comment by k88hudson** [comment]
don't mean to derail but I feel like it has to be said

**Comment by Pomax** [comment]
@k88hudson see [my comment](https://github.com/MozillaFoundation/MoFo-Engineering-Handbook/pull/23#issuecomment-69815711) about why I'm putting us back on Express.

File: app/_posts/2014-12-03-the-stack.md (hunk 2 above)
Comment (thisandagain) [inline app/_posts/2014-12-03-the-stack.md]: Seems odd to veto `joi` but not provide an alternative for object validation... JSON schema?

**Comment by thisandagain** [comment]
@Pomax Seems odd to veto `joi` but not specify an alternative for object validation... JSON schema? I agree with you re: smaller community and poor docs, but I'd like to see a plan for how we do router testing in Express and input validation.

**Comment by davidascher** [comment]
Suggest we split up this issue into a few: server-side routing framework,
testing framework, storage (although my own take is that's a tough one to
answer w/o knowledge of the specifics), object validation, etc.

**Comment by gvn** [comment]
FWIW this was originally intended to be our *front end* stack chapter, so yeah I agree splitting it up might be prudent. :+1:

**Comment by Pomax** [comment]
+1 on splitting this up

**Pull request closed without merging**